File: frontend/src/app/models/speaker.ts
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { SpeakersService } from '../deck-api/speakers.service';

export class Speaker {
    id: String;
    name: String;
    contact: String;
    title: String;
    bio: String;
    notes: String;
    imgs: SpeakerImages;
    participations: SpeakerParticipation[];
}

export class SpeakerImages {
    internal: String;
    speaker: String;
    company: String;
}

export class SpeakerParticipation {
    event: Number;
    member: String;
    status: String;
    communications: String[];
    subscribers: String[];
    feedback: String;
    flights: String[];
    room: SpeakerParticipationRoom;
}

export class SpeakerParticipationRoom {
    type: String;
    cost: Number;
    notes: String;
}

export class SpeakerParticipationValidStatusStep {
    next: String;
    step: Number;
}

export class SpeakerParticipationValidStatusSteps {
    steps: SpeakerParticipationValidStatusStep[];
}

export function GetParticipation(speaker: Speaker, event: Number): SpeakerParticipation {
    for (const participation of speaker.participations) {
        if (participation.event === event) {
            return participation;
        }
    }

    return null;
}

export class EditSpeakerForm {

    form: FormGroup;

    constructor(speaker: Speaker) {
        this.form = new FormGroup({
            bio: new FormControl(speaker.bio ? speaker.bio : '', [Validators.required, Validators.minLength(1)]),
            name: new FormControl(speaker.name, [Validators.required, Validators.minLength(1)]),
            notes: new FormControl(speaker.notes, [Validators.required, Validators.minLength(1)]),
            title: new FormControl(speaker.title, [Validators.required, Validators.minLength(1)]),
        });
    }

    value() {
        return this.form.value;
    }

    valid() {
        return this.form.valid;
    }

}
export class AddSpeakerForm {

    form: FormGroup;

    constructor() {
        this.form = new FormGroup({
            name: new FormControl('', [Validators.required, Validators.minLength(1)]),
            bio: new FormControl('', []),
            title: new FormControl('', [Validators.required, Validators.minLength(1)]),
        });
    }

    value() {
        return this.form.value;
    }

    valid() {
        return this.form.valid;
    }
}

export class EditSpeakerParticipationForm {

    form: FormGroup;

    constructor(participation?: SpeakerParticipation) {
        this.form = new FormGroup({
            feedback: new FormControl(participation ? participation.feedback : '', []),
            member: new FormControl(participation ? participation.member : '', [Validators.required, Validators.minLength(1)]),
            room: new FormGroup({
                cost: new FormControl(participation && participation.room ? participation.room.cost : 0, [Validators.required, Validators.min(0)]),
                notes: new FormControl(participation && participation.room ? participation.room.notes : '', []),
                type: new FormControl(participation && participation.room ? participation.room.type : '', []),
            })
        });
    }

    value() {
        return this.form.value;
    }

    valid() {
        return this.form.valid;
    }
}

export class EditSpeakerImageForm {

    form: FormGroup;
    file: File;

    constructor() {
        this.form = new FormGroup({
            image: new FormControl(null, [Validators.required]),
        });
    }

    value() {
        return this.file;
    }

    valid() {
        return this.form.valid;
    }

    set(file: File) {
        this.file = file;
    }
}

export class EditSpeakerParticipationStatusForm {

    form: FormGroup;
    options: SpeakerParticipationValidStatusSteps;
    speaker: Speaker;
    label: String;

    constructor(private speakersService: SpeakersService, speaker: Speaker, participation?: SpeakerParticipation) {
        this.form = new FormGroup({
            step: new FormControl('', [Validators.required, Validators.minLength(1)]),
        });

        this.speaker = speaker;

        this.updateOptions(participation);
    }

    updateOptions(participation: SpeakerParticipation) {
        if (participation === undefined || participation === null) { return; }

        this.speakersService.getValidStatusSteps(`${this.speaker.id}`)
            .subscribe((steps: SpeakerParticipationValidStatusSteps) => {
                this.options = steps;
            }
            );
    }

    set(step: SpeakerParticipationValidStatusStep) {
        this.form.controls['step'].setValue(step.step);
        this.label = step.next;
    }

    valid() {
        return this.form.valid;
    }

    value() {
        return this.form.get('step').value;
    }
}

File: frontend/lib/models/speaker.dart
import 'package:frontend/main.dart';
import 'package:frontend/models/participation.dart';
import 'package:collection/collection.dart';

class Images {
  final String? company;
  final String? internal;
  final String? speaker;

  Images({this.company, this.internal, this.speaker});

  factory Images.fromJson(Map<String, dynamic> json) {
    return Images(
        company: json['company'],
        internal: json['internal'],
        speaker: json['speaker']);
  }

  Map<String, dynamic> toJson() =>
      {'company': company, 'internal': internal, 'speaker': speaker};
}

class Speaker {
  final String id;
  final String name;
  final String? title;
  final String? contact; //TODO: lazy load
  final String? bio;
  final String? notes;
  final Images? imgs;
  final List<SpeakerParticipation>? participations;
  final int? numParticipations;
  final int? lastParticipation;
  final ParticipationStatus? participationStatus;

  Speaker(
      {required this.id,
      required this.name,
      this.title,
      this.contact,
      this.bio,
      this.notes,
      this.imgs,
      this.participations,
      this.numParticipations,
      this.lastParticipation,
      this.participationStatus});

  factory Speaker.fromJson(Map<String, dynamic> json) {
    var participationsList = json['participations'] as List;
    return Speaker(
      id: json['id'],
      name: json['name'],
      title: json['title'],
      contact: json['contact'],
      bio: json['bio'],
      notes: json['notes'],
      imgs: Images.fromJson(json['imgs']),
      participations: participationsList
          .map((p) => SpeakerParticipation.fromJson(p))
          .toList(),
      numParticipations: participationsList.length,
      lastParticipation: participationsList.length > 0
          ? participationsList[participationsList.length - 1]['event']
          : null,
      participationStatus: participationsList.length > 0 &&
              participationsList[participationsList.length - 1]['event'] ==
                  App.localStorage.getInt("event")
          ? Participation.convert(
              participationsList[participationsList.length - 1]['status'])
          : ParticipationStatus.NO_STATUS,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'title': title,
        'contact': contact,
        'bio': bio,
        'notes': notes,
        'imgs': imgs?.toJson(),
        'participations':
            participations?.map((participation) => participation.toJson())
      };

  SpeakerParticipation? getParticipation(int event) {
    return this
        .participations!
        .firstWhereOrNull((element) => element.event == event);
  }

  SpeakerParticipation getLatestParticipation() {
    return this.participations!.firstWhereOrNull(
        (element) => element.event == this.lastParticipation)!;
  }

  bool operator ==(o) => o is Speaker && id == o.id;
  int get hashCode => id.hashCode;
}

class PublicImages {
  final String? company;
  final String? speaker;

  PublicImages({this.company, this.speaker});

  factory PublicImages.fromJson(Map<String, dynamic> json) {
    return PublicImages(company: json['company'], speaker: json['speaker']);
  }

  Map<String, dynamic> toJson() => {'company': company, 'speaker': speaker};
}

class PublicSpeaker {
  final String? id;
  final String? name;
  final String? title;
  final Images? imgs;

  PublicSpeaker({this.id, this.name, this.title, this.imgs});

  factory PublicSpeaker.fromJson(Map<String, dynamic> json) {
    return PublicSpeaker(
        id: json['id'],
        name: json['name'],
        title: json['title'],
        imgs: Images.fromJson(json['imgs']));
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'title': title,
        'imgs': imgs?.toJson(),
      };
}

class SpeakerLight {
  final String id;
  final String name;
  final Images speakerImages;
  final int? numParticipations;
  final int? lastParticipation;
  final ParticipationStatus participationStatus;

  SpeakerLight(
      {required this.id,
      required this.name,
      required this.speakerImages,
      this.numParticipations,
      this.lastParticipation,
      required this.participationStatus});

  factory SpeakerLight.fromJson(Map<String, dynamic> json) {
    var participations = json['participations'] as List;
    return SpeakerLight(
        id: json['id'],
        name: json['name'],
        speakerImages: Images.fromJson(json['imgs']),
        numParticipations: participations.length,
        lastParticipation: participations.length > 0
            ? participations[participations.length - 1]['event']
            : null,
        participationStatus: participations.length > 0 &&
                participations[participations.length - 1]['event'] ==
                    App.localStorage.getInt("event")
            ? Participation.convert(
                participations[participations.length - 1]['status'])
            : ParticipationStatus.NO_STATUS);
  }
}